packages/aws-cdk/lib/api/garbage-collection/garbage-collector.ts

```diff
@@ -178,14 +208,106 @@ export class GarbageCollector {
     });
     backgroundStackRefresh.start();
 
+    try {
+      if (this.garbageCollectS3Assets) {
+        await this.garbageCollectS3(sdk, activeAssets, backgroundStackRefresh);
+      }
+
+      if (this.garbageCollectEcrAssets) {
+        await this.garbageCollectEcr(sdk, activeAssets, backgroundStackRefresh);
+      }
+    } catch (err: any) {
+      throw new Error(err);
+    } finally {
+      backgroundStackRefresh.stop();
+    }
+  }
+
+  /**
+   * Perform garbage collection on ECR assets
+   */
+  public async garbageCollectEcr(sdk: ISDK, activeAssets: ActiveAssetCache, backgroundStackRefresh: BackgroundStackRefresh) {
+    const ecr = sdk.ecr();
+    const repo = await this.bootstrapRepositoryName(sdk, this.bootstrapStackName);
+    const numImages = await this.numImagesInRepo(ecr, repo);
+    const printer = new ProgressPrinter(numImages, 1000);
+
+    debug(`Found bootstrap repo ${repo} with ${numImages} images`);
+
+    try {
+      // const batches = 1;
+      const batchSize = 1000;
+      const currentTime = Date.now();
+      const graceDays = this.props.rollbackBufferDays;
+
+      debug(`Parsing through ${numImages} images in batches`);
+
+      for await (const batch of this.readRepoInBatches(ecr, repo, batchSize, currentTime)) {
+        await backgroundStackRefresh.noOlderThan(600_000); // 10 mins
+        printer.start();
+
+        const { included: isolated, excluded: notIsolated } = partition(batch, asset => !asset.tags.some(t => activeAssets.contains(t)));
+
+        debug(`${isolated.length} isolated images`);
+        debug(`${notIsolated.length} not isolated images`);
+        debug(`${batch.length} images total`);
+
+        let deletables: ImageAsset[] = isolated;
+        let taggables: ImageAsset[] = [];
+        let untaggables: ImageAsset[] = [];
+
+        if (graceDays > 0) {
+          debug('Filtering out images that are not old enough to delete');
+
+          // We delete images that are not referenced in ActiveAssets and have the Isolated Tag with a date
+          // earlier than the current time - grace period.
+          deletables = isolated.filter(img => img.isolatedTagBefore(new Date(currentTime - (graceDays * DAY))));
```

Comment:
Do tags even work this way in an ECR repository? 

Doesn't a tag only ever refer to exactly 1 image?

Reply:
I think the difference here is:

* In S3, tags are nonfunctional metadata
* In ECR (Docker), tags are a functional part of how you address the image

Reply:
You can AWS-tag the repository itself... but it has limits key <128 chars, value <256 chars

Reply:
i believe that the tags work the way i have written. you can have [multiple tags](https://stackoverflow.com/questions/75417591/can-i-apply-another-tag-to-an-existing-image-in-the-ecr-web-interface) per image. tag [immutability](https://docs.aws.amazon.com/AmazonECR/latest/userguide/image-tag-mutability.html) shouldn't apply here - we are not overwriting any existing tags. 

you could argue that it looks weird to tag our isolated images this way -- but i think this is at least _doable_.

Reply:
Here's an example of what it looks like in the console:

<img width="1339" alt="Screenshot 2024-10-22 at 4 29 01 PM" src="https://github.com/user-attachments/assets/605f8c2a-85ab-40c6-914d-b8152a5e512e">

Reply:
> > Doesn't a tag only ever refer to exactly 1 image?

> you can have [multiple tags](https://stackoverflow.com/questions/75417591/can-i-apply-another-tag-to-an-existing-image-in-the-ecr-web-interface) per image.

That's not what I said though. Can you have 2 images with the same tag? 

If not, how are you going to tag multiple images for garbage collection? Are you guaranteeing that the timestamps are all unique?

I guess conceivably it might be doable if you use JavaScript timestamps with millisecond resolution and increment them for each; we'd need to tag more than 1000 images/second for that to become inaccurate enough to care about it (and even then... meh).

Even given this... I'm not sure I'm comfortable putting this metadata in a load-bearing location. I *guess* we can try it and see if it breaks things, or puts undue load on ECR

Reply:
There can be at most 50 AWS tags per repo, looks like.... that's not great.

Also, there can only be at most 1000 Docker tags per repo, which means at most 500 images per repo (since every image could have an asset hash tag and an isolated tag). And once we tag a bunch of images, we can't add new ones.

Reply:
We've decided offline to go with the Docker tags method as repository tags are limited in number (~50 metadata tags per resource). The issue with Docker tags is ensuring that each tag is unique, and we will do that via timestamps + increments.